samples/nrf_desktop/src/modules/config.c

```diff
@@ -0,0 +1,231 @@
+	 * We can apply values to application.
+	 */
+	for (u8_t id = 0; id < CONFIG_EVENT_ID_MAX; id++) {
+		if (!loaded_data[id]) {
+			continue;
+		}
+
+		EVENT_SUBMIT(loaded_data[id]);
+		loaded_data[id] = NULL;
+	}
+
+	return 0;
+}
+
+static void update_config(const u8_t config_id, u8_t *data, size_t size)
+{
+	char key[KEY_LEN];
+	int err = snprintk(key, sizeof(key), MODULE_NAME "/%u", config_id);
+	__ASSERT_NO_MSG(err < sizeof(key));
+
+	if (err < 0) {
+		LOG_ERR("Creating key failed");
+	} else {
+		LOG_INF("Store %s", log_strdup(key));
+		err = settings_save_one(key, data, size);
+		if (err) {
+			LOG_ERR("Cannot store err %d", err);
+		}
+	}
+}
+
+static int enable_settings(void)
+{
+	int err = 0;
+
+	if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE)) {
+		static struct settings_handler sh = {
+			.name = MODULE_NAME,
+			.h_set = settings_set,
+			.h_commit = commit,
+		};
+
+		err = settings_register(&sh);
+		if (err) {
+			LOG_ERR("Cannot register settings handler");
+			goto error;
+		}
+	}
+
+	/* This module loads settings for all application modules */
+	err = settings_load();
+	if (err) {
+		LOG_ERR("Cannot load settings");
+		goto error;
+	}
+
+	LOG_INF("Settings loaded");
+
+error:
+	return err;
+}
+
+static void init(void)
+{
+	int err = enable_settings();
+
+	if (err) {
+		module_set_state(MODULE_STATE_ERROR);
+	} else {
+		module_set_state(MODULE_STATE_READY);
+	}
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE)) {
+		if (is_config_event(eh)) {
+			struct config_event *event = cast_config_event(eh);
+
+			/* Accept only events coming from transport. Do not
+			 * write already stored information.
+			 */
+			if (event->store_needed) {
+				update_config(event->id, event->data,
+					      sizeof(event->data));
+
+			}
+
+			return false;
+		}
+	}
+
+	if (is_module_state_event(eh)) {
+		const void * const req_modules[] = {
+			MODULE_ID(main),
+#if CONFIG_DESKTOP_HIDS_ENABLE
+			MODULE_ID(hids),
+#endif
+#if CONFIG_DESKTOP_BAS_ENABLE
+			MODULE_ID(bas),
+#endif
+#if CONFIG_DESKTOP_BLE_ADVERTISING_ENABLE
+			MODULE_ID(ble_adv),
+#endif
+#if CONFIG_DESKTOP_MOTION_OPTICAL_ENABLE
+			MODULE_ID(motion),
+#endif
+#if CONFIG_DESKTOP_BLE_SCANNING_ENABLE
+			MODULE_ID(ble_scan),
+#endif
+		};
+
+		struct module_state_event *event = cast_module_state_event(eh);
+
+		/* Settings need to be loaded after GATT services are setup,
+		 * otherwise the values stored in flash will not be written
+		 * to GATT database.
+		 */
+		static u32_t req_state;
+		static_assert(ARRAY_SIZE(req_modules) < (8 * sizeof(req_state)),
+			      "Array size bigger than number of bits");
+
+		for (size_t i = 0; i < ARRAY_SIZE(req_modules); i++) {
+			if (check_state(event, req_modules[i], MODULE_STATE_READY)) {
+				unsigned int flag = BIT(i);
+
+				/* Catch double initialization of any module */
+				__ASSERT_NO_MSG((req_state & flag) == 0);
```

Comment:
`motion` can be suspended and resumed so assert will crash on double ready for it